File: cpp/ProxyDemo.cpp
#include <stdio.h>
#include <string>
#include <iostream>
using namespace std;

class Server{
	public:
		virtual ~Server(){}
		virtual void service(const string& req, string& rsp) = 0;
};

class HttpServer: public Server{
	public:
		void service(const string& req, string& rsp) {
			cout<<"handle req"<<req<<endl;;
		}
};


class HttpServerProxy: public Server{
	public:
		void service(const string& req, string& rsp)	{
			if (req.find("sbkun") !=  string::npos) {
				cout<<"not handle req"<<req<<endl;
			}
			else{
				server_.service(req, rsp);
			}
		}
	private:
		HttpServer server_;
};


int main(int argc, char* argv[])
{
	HttpServerProxy sp;
	string rsp;
	sp.service("sbkun", rsp);
	sp.service("haha", rsp);
	return 0;
}

File: java/structural/ProxyDemo.java
package structural;

class HttpService {
    void service(String req, String rsp) {
        System.out.println("handle req " + req);
    }
}

class HttpServerProxy extends HttpService {
    void service(String req, String rsp) {
        if (req.indexOf("sbkun") == -1) {
            super.service(req, rsp);
        } else {
            System.out.println("not handle " + req);
        }
    }
}

public class ProxyDemo {
    public static void main(String args[]) {
        HttpServerProxy hp = new HttpServerProxy();
        String rsp = "";
        hp.service("sbkun", rsp);
        hp.service("haha", rsp);
    }
}
